### ISSUE TITLE
Chrome breaks layout if system language is Chinese

### ISSUE THREAD

**Issue by cyent** [squidfunk/mkdocs-material#911]
## Description

Only chrome meet the bug，safari and firefox are correct.

the case include my "mkdocs serve" and your "https://squidfunk.github.io/mkdocs-material/"

## this is chrome:
![image](https://user-images.githubusercontent.com/1782125/47857443-954f6f00-de24-11e8-8dea-35286d7c173c.png)

## this is firefox:
![image](https://user-images.githubusercontent.com/1782125/47857479-a6987b80-de24-11e8-84cc-5ce1e2ceb1af.png)

### Package versions

* Python: `2.7.10`
* MkDocs: `1.0.4`
* Material: `3.0.6`

### Project configuration

``` yaml
site_name: My Docs
theme:
  name: material
```

### System information

* OS: macos 10.13.4
* Browser: chrome

**Comment by cyent** [comment]
I mean, in the chrome, some size display correct, some size have problem. Now, I take 3 screenshots in another windows computer，the chrome version is 70.0.3538.67.

Now, let me visit https://squidfunk.github.io/mkdocs-material/getting-started/

## screenshot 1：the same as mobilephone，it is correct
![image](https://user-images.githubusercontent.com/1782125/47947604-d9449000-df5a-11e8-9c58-aabbb7bb72ea.png)

## screenshot 2：the size is bigger than screenshot 1，it is correct
![image](https://user-images.githubusercontent.com/1782125/47947608-dd70ad80-df5a-11e8-9891-553103d4bd38.png)

## screenshot 3：the size is bigger than screenshot 2，it has problem：I can't see the Right sidebar fully
![image](https://user-images.githubusercontent.com/1782125/47947607-dcd81700-df5a-11e8-8816-f083e1b3715a.png)

**Comment by squidfunk** [comment]
I'm afraid I cannot reproduce this. Is this the latest Chrome version? Somebody experiencing the same problem?

**Comment by cyent** [comment]
eh, eh, actually, the problem appear at least 1.5 years（at that time, I just learn mkdocs-material）...

**Comment by squidfunk** [comment]
As said before, I'm not able to reproduce it on Chrome macOS. It's very strange that this is happening for you. Maybe somebody else has experienced the same, we need a way to reproduce it.

**Comment by cyent** [comment]
I don't know how to reproduce it , I try win10 chrome、macOS chrome，they are different computers. So it's very strange. I will try in other people's computer again, and tell you.

**Comment by squidfunk** [comment]
Still (macOS, latest Chrome): 

<img width="1412" alt="bildschirmfoto 2018-11-07 um 10 18 47" src="https://user-images.githubusercontent.com/932156/48121935-92b2a680-e276-11e8-9f55-ea48fb977439.png">

**Comment by martinbira** [comment]
Windows 10, latest Chrome here @ 1300x368
![material-mkdocs-1300x368](https://user-images.githubusercontent.com/22232681/48122768-9e06d180-e278-11e8-94f1-6b20bf444766.png)

**Comment by cyent** [comment]
I reinstall the latest chrome, but the problem still exists, these days I will try a few more times in more other peoples.

**Comment by cyent** [comment]
I can reproduce it, it is the OS Language（not chrome Language）.

**When the os language is chinese simplified, it will reproduce the problem**
When the os language is english, it is very ok.

example macos
![image](https://user-images.githubusercontent.com/1782125/48243947-eb707380-e41d-11e8-939d-a0de7a010264.png)
If you set it like this, you can reproduce it. （I haven't tried it on Windows yet）

**Comment by walkccc** [comment]
In my friend's MacBook Pro 13" (2013 mid), there's the same issue as following (also in chrome browser and Traditional Chinese; however, it's perfect in Safari).

![issue](https://user-images.githubusercontent.com/14782909/48284964-13062100-e49c-11e8-9eaf-fd4b5a7df668.jpeg)

**Comment by squidfunk** [comment]
Closing, as its still not reproducible. If somebody manages to debug (and maybe fix it) you are most welcome to submit a PR.

**Comment by smancill** [comment]
I can reproduce the problem with Chrome on macOS:

<img width="1285" alt="captura de pantalla 2019-02-04 a la s 04 43 52" src="https://user-images.githubusercontent.com/238528/52227741-26543d00-288f-11e9-8515-b0fc72cafac8.png">

The `margin-left` of `md-sidebar--secondary` is set to the wrong value.

**Comment by squidfunk** [comment]
@smancill - did you use browser zoom? What's your base font size?

**Comment by smancill** [comment]
The base font size is the browser default (medium, 16pt).

And no zoom (even though the above screenshot in responsive mode was at 75%).

Here is another one, this time at 100%:

<img width="1437" alt="captura de pantalla 2019-02-04 a la s 16 36 32" src="https://user-images.githubusercontent.com/238528/52232386-235f4980-289b-11e9-90f7-c4d54808885d.png">


By the way, my system language is Spanish but I changed it to English and the problem was still present.

**Comment by smancill** [comment]
Ah, I found the actual problem.

There is a preference in Chrome to set the **minimum font size**. I have it set to 12px (no idea if it is the default or if I changed at some point).

Setting that to 10px fixes the issues.

**Comment by squidfunk** [comment]
Awesome, thanks! Setting the minimum font size to `12px` I can finally reproduce it! For me the value was set to `6px` and I never tweaked it, so it should not be the default I guess. I will see if we can do something about it, but I would consider it an edge case.

**Comment by smancill** [comment]
I actually found the issue with Chrome. There is a hidden setting in Chrome that I cannot change from the preferences panel UI:

```
$ jq '.' ~/Library/Application\ Support/Google/Chrome/Default/Preferences | grep font_size
      "default_fixed_font_size": 13,
      "default_font_size": 16,
      "minimum_font_size": 10,
      "minimum_logical_font_size": 10,
```

I had both `minimum_font_size` and `minimum_logical_font_size` set to 12, but from the UI I can only change `minimum_font_size` to 10, and `minimum_logical_font_size` was still at 12 and the problem still persisted.

I had to quit Chrome, edit the preferences file manually to set `minimum_logical_font_size` to 10 and finally the layout was correct.

**Comment by squidfunk** [comment]
@cyent and @walkccc - can you please look into your minimum font size settings and verify whether the issue persists if you set it to `10px`?

**Comment by walkccc** [comment]
I think the issue still persists. Changing the `font-size` didn't affect the layout.

- `font-size: 10px`

    <img width="1439" alt="10" src="https://user-images.githubusercontent.com/14782909/52530113-2e065c80-2d3b-11e9-928c-06f14d074cf3.png">

- `font-size: 12px`

    <img width="1442" alt="12" src="https://user-images.githubusercontent.com/14782909/52530110-1202bb00-2d3b-11e9-9e75-f71301100d26.png">

**Comment by squidfunk** [comment]
@walkccc you have to set the font size within Chrome settings, not DevTools. However @smancill research led to [an unresolved issue](
https://bugs.chromium.org/p/chromium/issues/detail?id=36429) I found which may explain why the issue for some users only appears for Chinese:

> By default, chrome sets the minimum font size as 1px for most UI locale, but for some special UI locale like Simp/Trad Chinese, Japanese, Korea and Thai, Chrome uses 10px/12px as default minimum font size. That is why Chrome does not display font which size is less than 12px in Chinese UI.

**Comment by walkccc** [comment]
### OS language = Trad Chinese.

- Very small
    <img width="1440" alt="2019-02-10 3 57 40" src="https://user-images.githubusercontent.com/14782909/52531217-c6f1a380-2d4c-11e9-9f9a-1fb3f804f6b4.png">

- Small
    <img width="1440" alt="2019-02-10 3 58 13" src="https://user-images.githubusercontent.com/14782909/52531218-c78a3a00-2d4c-11e9-8faf-41b4a8930acc.png">

- Medium (Recommended)
    <img width="1440" alt="2019-02-10 3 58 23" src="https://user-images.githubusercontent.com/14782909/52531219-c78a3a00-2d4c-11e9-9a97-ad8c2a1d5410.png">

- Big
    <img width="1440" alt="2019-02-10 3 58 34" src="https://user-images.githubusercontent.com/14782909/52531220-c78a3a00-2d4c-11e9-8edd-a9a2f66d2d42.png">

- Very big
    <img width="1440" alt="2019-02-10 3 58 42" src="https://user-images.githubusercontent.com/14782909/52531221-c822d080-2d4c-11e9-9c9d-7af7b2a765cd.png">

**Comment by squidfunk** [comment]
Thanks, I will see if we can fix it somehow.

**Comment by squidfunk** [comment]
Only Chinese should be affected, as Material sets the root font size to 10 for an efficient usage of `rem` (see [here](https://bugs.chromium.org/p/chromium/issues/detail?id=16875)):

> Set the default minimum font size webkit pref to 12 for Chinese and 10 for ja,ko,ar and th.

**Comment by squidfunk** [comment]
So, the problem is definitely triggered by the fact that the minimal font size for Chinese cannot be smaller than `12`. The issue was reported 9 (!!!) years ago and still remains present in Chromium and Chrome, see [here](https://bugs.chromium.org/p/chromium/issues/detail?id=36429) and [here](https://bugs.chromium.org/p/chromium/issues/detail?id=16875).

This year-standing "fix" in Chrome greatly hinders accessibility. Material puts a huge effort in being as accessible as possible by defining every dimension in `rem` and Media Queries in `em`. If a user has set a bigger base font size (like it would be practical for elderly people), the whole page will just appear as if it is zoomed. This ensures that the layout always looks as intended. The problem is that this breaks all `rem` values, as the factor is now 1.2 and not 1.0. However, the breakpoints defined in the Media Queries stay at their positions, which is the reason why the sidebar is way of the screen.

We could try to provide a "good enough" experience by somehow working around this issue in the CSS or base all computations on a base font size of 12 (and not 10) which would require some rewriting of the SASS code (probably using a helper function). The problem will persist if a user set the minimal font size to a value larger than 12. It is not fixable with the current architecture without sacrificing accessibility.

**Comment by masterstevelu** [comment]
> I'm afraid I cannot reproduce this. Is this the latest Chrome version? Somebody experiencing the same problem?



> So, the problem is definitely triggered by the fact that the _minimal font size_ for Chinese cannot be smaller than `12`. The issue was reported 9 (!!!) years ago and still remains present in Chromium and Chrome, see [here](https://bugs.chromium.org/p/chromium/issues/detail?id=36429) and [here](https://bugs.chromium.org/p/chromium/issues/detail?id=16875).
> 
> This year-standing "fix" in Chrome greatly hinders accessibility. Material puts a huge effort in being as accessible as possible by defining every dimension in `rem` or `em` and Media Queries in `em`. If a user has set a bigger _base font size_ (like it would be practical for elderly people), the whole page will just appear as if it is zoomed. This ensures that the layout always looks as intended. The problem is that a _minimal font size_ of `12` breaks all `rem` values, as the factor is now `1.2` and not `1.0`. However, the _minimal font size_ does not seem to impact the _base font size_ for the Media Query computations, as they stay at their positions which is the reason why the sidebar is way of the screen (Media Query now triggers too early).
> 
> We could try to provide a "good enough" experience by somehow working around this issue in the CSS or base all computations on a _base font size_ of `12` (and not `10`) which would require some rewriting of the SASS code (probably using a helper function). The problem will persist if a user sets the _minimal font size_ to a value larger than `12`. It is not fixable with the current architecture without sacrificing accessibility.
> 
> I really do not want to lock out the whole Chinese community and I'm really shocked that such a fundamental issue remains unsolved for such a long time, impacting a huge part of the world's population. Refactoring all `rem` values to be based on a factor of `1.2` to catch the _minimal font size_ for Chinese will lead to regressions for users which applied customizations in `rem`, as their calculations are now off. This is a tricky issue.

I encountered the same problem today. And thank you for your efforts on fixing this bug!

**Comment by squidfunk** [comment]
Looks like I have good news – I just pushed a potential workaround in 20074843 and 946f4f1 to `master`. The gist:

- Move all `rem` calculations to a new helper function `px2rem` and define all values in `px`. The helper function will take the `px` value and return the appropriate `rem` value using the base font-size.
- Double the base font-size on the `html` element from 62.5% (= `10px`) to 125% (= `20px`) and adjust `px2rem` accordingly. This works around the problem that Chrome will internally change `10px` to `12px` and should work for a minimal font size of up to `20px`.
- Reset the base font size to `10px` by defining `0.5rem` on the `body`, so that implicit inline spacing is as it was before.

My testing shows that the issue is indeed fixed. The site looks exactly the same when I set the minimal font size to `12px` so if my theory is correct it should now work for Chinese systems.

**The change is already deployed to the live site, so everyone experiencing the problem please visit https://squidfunk.github.io/mkdocs-material/ to check if the problem is gone**

Before:

<img width="700" alt="bildschirmfoto 2019-02-12 um 19 15 27" src="https://user-images.githubusercontent.com/932156/52659514-b2a0d880-2efd-11e9-8c81-da9d61274c3d.png">

After:

<img width="700" alt="bildschirmfoto 2019-02-12 um 19 15 29" src="https://user-images.githubusercontent.com/932156/52659524-b7658c80-2efd-11e9-9424-2d590f049569.png">

If we come to the conclusion that the issue is fixed, the next steps are:

1. I would kindly ask somebody to review commit 20074843 which changes `0.4rem` to `px2rem(4px)` etc. I double checked and it seems that I made no mistakes but there are A LOT of changes, so it's always better to double check.
2. I will issue a major release, i.e. version 4, because this change has the potential to break some projects using customization. If a project uses `rem` values, the resulting values are now twice as big. From my experience only very few people use `rem`, so it shouldn't break a lot of projects, but it still has the potential to do so.

What. A. Journey.

I think we're all much wiser now and it seems that we're actually the first discovering the problem of minimal font size in respect to `rem` values.

**Comment by squidfunk** [comment]
@masterstevelu @walkccc @cyent @smancill – could someone of you please verify, whether the issue is now gone when visiting the official Material docs?

https://squidfunk.github.io/mkdocs-material/

**Comment by masterstevelu** [comment]
> @masterstevelu @walkccc @cyent @smancill – could someone of you please verify, whether the issue is now gone when visiting the official Material docs?
> 
> https://squidfunk.github.io/mkdocs-material/

The bug is fixed on my Chrome. Thannk you!

**Comment by masterstevelu** [comment]
@squidfunk Hi Martin, when can I use your newest build if I install it using pip?

**Comment by walkccc** [comment]
It's gone!

**Comment by squidfunk** [comment]
Awesome! As described above, I would like somehow to review 200748430 before I will issue a new major release. We need to make sure that all `rem` units were correctly converted to `px`:

- `0.1rem` => `px2rem(1px)`
- `0.9rem` => `px2rem(9px)`
- `1rem` => `px2rem(10px)`
- `1.1rem` => `px2rem(11px)`
- `24.2rem` => `px2rem(242px)`
- etc.

If we missed something, it will break some projects down the line. Better safe than sorry. I will prepare the release notes and upgrade instructions, so we can issue a new release  as soon as we're sure we won't break anything.

**Comment by squidfunk** [comment]
Released as part of 4.0.0! I want to thank you all for your effort on helping fix this issue. Material should now finally look as it it should in China :-)
